src/log-store/src/kafka/index.rs

```diff
@@ -0,0 +1,286 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::VecDeque;
+use std::ops::Range;
+
+use itertools::Itertools;
+use store_api::logstore::EntryId;
+
+/// An iterator over WAL (Write-Ahead Log) entries index for a region.
+pub trait RegionWalIndexIterator: Send + Sync {
+    // Returns the size of the next batch.
+    fn next_batch_size(&self) -> u64;
+
+    // Peeks the next EntryId without advancing the iterator.
+    fn peek(&self) -> Option<EntryId>;
+
+    // Advances the iterator and returns the next EntryId.
+    fn next(&mut self) -> Option<EntryId>;
+}
+
+pub struct RegionWalRange {
+    next_entry_id: EntryId,
+    end_entry_id: EntryId,
+}
+
+impl RegionWalRange {
+    pub fn new(range: Range<EntryId>) -> Self {
+        Self {
+            next_entry_id: range.start,
+            end_entry_id: range.end,
+        }
+    }
+}
+
+impl RegionWalIndexIterator for RegionWalRange {
+    fn next_batch_size(&self) -> u64 {
+        if self.next_entry_id < self.end_entry_id {
+            self.end_entry_id
+                .checked_sub(self.next_entry_id)
+                .unwrap_or_default()
+        } else {
+            0
+        }
+    }
+
+    fn peek(&self) -> Option<EntryId> {
+        if self.next_entry_id < self.end_entry_id {
+            Some(self.next_entry_id)
+        } else {
+            None
+        }
+    }
+
+    fn next(&mut self) -> Option<EntryId> {
+        if self.next_entry_id < self.end_entry_id {
+            let next = self.next_entry_id;
+            self.next_entry_id += 1;
+            Some(next)
+        } else {
+            None
+        }
+    }
+}
+
+pub struct RegionWalVecIndex {
+    index: VecDeque<EntryId>,
+}
+
+impl RegionWalVecIndex {
+    pub fn new<I: IntoIterator<Item = EntryId>>(index: I) -> Self {
+        Self {
+            index: index.into_iter().collect::<VecDeque<_>>(),
+        }
+    }
+}
+
+impl RegionWalIndexIterator for RegionWalVecIndex {
+    fn next_batch_size(&self) -> u64 {
+        if self.index.is_empty() {
+            0
+        } else {
+            let mut count = 1;
+            for (prev, next) in self.index.iter().tuple_windows() {
+                if prev + 1 == *next {
+                    count += 1;
+                } else {
+                    break;
+                }
+            }
+
+            count
+        }
+    }
```

Comment:
Perhaps we can make the return value of next_batch_size merge some overly small ranges according to a page (a page might be the most economical size for Kafka client read operations).

This way, we combine N read operations into one read operation. Although the I/O bandwidth has increased slightly, it significantly reduces latency.